`SqliteDosStorage`: Catch locked database when setting global variable

sphuber opened the pull request:
Fixes #6532 

When running multiple processes simultaneously, even for a single runner whether local or daemonized, `sqlalchemy.exc.OperationError` exceptions would be thrown. They were due to the `set_process_state_change_timestamp` utility function which updates a row in the `DbSetting` table. The function is called each time a `Process` changes state.

This would cause problems for the `SqliteDosStorage` as that uses SQLite as the database which usually allows at most one writer to proceed concurrently. Luckily, this is the only place where this would crop up, as this is the only place in the database where different processes can touch the same row. SQLite is being operated in WAL mode in which case concurrent operations are supported as long as they touch different pages. This translates roughly to it being ok for different processes writing to the database as long as they are not touching the same rows. Since in AiiDA each process can only ever be run by a single runner, there should never be runners that are writing to the same row in a table.

The one exception is the writing of the `'process|state_change|.*` key in the `DbSetting` table during each process state change. This value is only used by `verdi process list` where it is displayed at the end, in order to give the user a sense of whether any processes are actually being run.

In that sense, it is not quite critical that every call is successful. Therefore, the `SqliteDosStorage` overrides the `set_global_variable` method to catch the `OperationalError` so that the failure can simply be logged and the process is not excepted. Since the reason for the failure is the fact that another process is already updating the timestamp, it is not critical if another writing is skipped. It will simply be updated again at a later stage.

sphuber commented on src/aiida/storage/sqlite_dos/backend.py:
```diff
@@ -313,6 +313,16 @@ def nodes(self):
+
+        try:
+            super().set_global_variable(key, value, description, overwrite)
+        except OperationalError:
```
Ideally, I would try to make sure the error here is really due to a locked database, but I am not sure that is possible in a deterministic way apart from parsing the error message, but that feels too fragile.

sphuber commented:
> I don't have the full context so sorry if this is a naive question.
> In your solution, it seems you'll ignore error from any global variables being set. Are there any others besides the process_state_change timestamp that are being set via `set_global_variable`? The current solution seems quite dangerous to me.

It is a fair question, and it could indeed be dangerous but for now setting the process state change timestamp is indeed the only use in the entire code base for using this method and always has been. That is why I think it is fine to work around the problem in this way for the SQLite storage plugin. Of course we would have to be weary of this in case we ever start using the settings for different things where each mutation is critical, but we haven't touched this in all of AiiDA's existence, so I think it is safe to do this for now.

There is another potential problem that is more problematic though that we still have to investigate before merging this. We have to make sure that if this exception is raised and we ignore it, there is not a transaction ongoing that is then rolled back that might include other valuable mutations that we cannot afford to lose. I am not yet sure how to test this of what is actually happening. I was even thinking of just opening a separate (nested) transaction in the `SqliteStorage.set_global_variable` such that only that transaction is rolled back in case of an exception that we can then safely ignore.

giovannipizzi commented:
I think I agree with @danielhollas. Wouldn't it be better to put the try-except here, around this line?

https://github.com/aiidateam/aiida-core/blob/d73731f428b031ad4b9f68a4af2a008adc9b3290/src/aiida/engine/utils.py#L288

Otherwise in the future we will forget and have very nasty bugs, very hard to debug.

Also, I think the putting an explicit transaction around that line could be a good idea.
For the tests, I would then do those (a bit extensive, mostly to try to catch possible future changes in behavior):

- run 2 operations including setting the timestamp without any explicit transaction as the second one (the first one could be something else, the best I see is setting an extra of a node), ensure that they both work
- same as above, but with both commands within the same transaction, and check that both succeed
- same as above, but with both commands with two transactions, and check that both succeed

Then, the same 3 tests, but where we make the error fail (monkeypatch?) and check that in all 3 cases, the first of the two commands is committed while the second one not

sphuber commented:
> Wouldn't it be better to put the try-except here, around this line?

It would have to be a broad except though since this part of the code is storage plugin agnostic and so we don't know what exception could be thrown.

> Otherwise in the future we will forget and have very nasty bugs, very hard to debug.

I see your point, but as I mentioned, the settings table has not be used for anything other than this process state change since the beginning of AiiDA. I acknowledge that it _could_ hide bugs for the SQLite implementation in the future, but likewise, catching the exception in the agnostic part of the code could hide bugs in the implementation of other storage plugins.

But ok fair enough, no matter where we put it, it is a huge work around to hide a pretty nasty gaping hole in the SQLite implementation due to the limited concurrent writes. We are really getting extremely lucky that it seems that under most "normal" operations we are only hitting the problem in this one part of the code. But I am afraid that "luck" will be of short duration and pretty soon there will be other avenues found that hit this problem.

I will move the patch to the storage plugin agnostic code and try to work out the tests.

sphuber commented:
@giovannipizzi I changed the implementation and added the tests. As for the transaction, I have only done the two cases, with and without a transaction, and not the intermediate case of a separate transaction for each case. This is mostly because it is slightly tricky to implement as a parametrized test and I think this is a case that is essentially more or less covered by the other ones.

I also changed the signature of the `set_process_state_change_timestamp` utility to faciliate testing. It was taking a `Process` instance but only using its associated node. Now it simply takes the node. Think this is ok since it is unlikely to be used by users. Technically we could go through a deprecation pathway but don't feel it is worth the effort in this case.

giovannipizzi reviewed (approved):
Thanks! All good with me and I agree with your comments! I have the gut feeling that with this, most of the concurrency issues are solved for 99% of the cases that are relevant for SQLite - let's see! :-)
